ieee802154/submac: fix initialization code

jia200x opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
This PR does essentially 3 things:
- Move some of the initialization code of netdev_ieee802154_submac from `netdev_submac_ieee802154_init` to the actual `_init` function of `netdev_driver_t`. This is required in order to make sure the device is initialized before setting netdev stuff.
- Fix the default SubGHz page
- Call `confirm_set_trx_state` after setting the state to RX_ON during initialization. In some cases, the radio was not ready to do a state transition because `request_set_trx_state` didn't finish.


<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
- Ensure hal dependent radios (cc2538, nrf52840) still work with `gnrc_networking` and `tests/ieee802154_hal`
<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references
None so far
<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

jeandudey commented on drivers/netdev_ieee802154_submac/netdev_ieee802154_submac.c:
```diff
@@ -291,8 +291,24 @@ static int _init(netdev_t *netdev)
+    /* This function already sets the PAN ID to the default one */
+    netdev_ieee802154_reset(netdev_ieee802154);
+
+    uint16_t chan = CONFIG_IEEE802154_DEFAULT_CHANNEL;
+    int16_t tx_power = CONFIG_IEEE802154_DEFAULT_TXPOWER;
```
```suggestion
    uint16_t chan = submac->channel_num;
    int16_t tx_power = submac->tx_pow;
```

Maybe get it from the submac values? so we don't lose sync between netdev and submac in case a change is made to `ieee802154_submac_init`.

jia200x replied:
you are right, we should definitely do that.

jeandudey commented on sys/net/link_layer/ieee802154/submac.c:
```diff
@@ -374,14 +374,15 @@ int ieee802154_submac_init(ieee802154_submac_t *submac, const network_uint16_t *
     ieee802154_phy_conf_t conf =
     { .phy_mode = submac->phy_mode,
       .channel = CONFIG_IEEE802154_DEFAULT_CHANNEL,
-      .page = CONFIG_IEEE802154_DEFAULT_CHANNEL,
+      .page = CONFIG_IEEE802154_DEFAULT_SUBGHZ_PAGE,
       .pow = CONFIG_IEEE802154_DEFAULT_TXPOWER };
```
Currently a different channel page is saved to `submac->channel_page` (and `channel_num` too) when 2.4 GHz or Sub-1-GHz is supported, maybe we can use these values there?

jia200x replied:
yes, of course. I completely miss that one

jia200x commented:
done!

jia200x commented on sys/net/link_layer/ieee802154/submac.c:
```diff
@@ -382,6 +382,7 @@ int ieee802154_submac_init(ieee802154_submac_t *submac, const network_uint16_t *
                                               CONFIG_IEEE802154_CCA_THRESH_DEFAULT) >= 0);
 
     ieee802154_radio_request_set_trx_state(dev, IEEE802154_TRX_STATE_RX_ON);
+    while(ieee802154_radio_confirm_set_trx_state(dev) == -EAGAIN) {};
```
yes, this must be fixed. I will piggyback some changes

jia200x commented:
I added some commits that improve error checks.
I also added assertions in cases where it's not possible to recover. This would indicate faulty hardware or that something really got wrong.

benpicco commented on drivers/netdev_ieee802154_submac/netdev_ieee802154_submac.c:
```diff
@@ -291,8 +291,28 @@ static int _init(netdev_t *netdev)
-    ieee802154_submac_init(submac, (network_uint16_t*) netdev_ieee802154->short_addr, (eui64_t*) netdev_ieee802154->long_addr);
+    netdev_ieee802154_setup(netdev_ieee802154);
+
+    int res = ieee802154_submac_init(submac, (network_uint16_t*) netdev_ieee802154->short_addr, (eui64_t*) netdev_ieee802154->long_addr);
```
```suggestion
    int res = ieee802154_submac_init(submac,
                                     (network_uint16_t*) netdev_ieee802154->short_addr,
                                     (eui64_t*) netdev_ieee802154->long_addr);
```

jia200x replied:
done!

benpicco commented on sys/net/link_layer/ieee802154/submac.c:
```diff
@@ -373,17 +391,18 @@ int ieee802154_submac_init(ieee802154_submac_t *submac, const network_uint16_t *
 
-    ieee802154_radio_request_set_trx_state(dev, IEEE802154_TRX_STATE_RX_ON);
+    _req_set_trx_state_wait_busy(dev, IEEE802154_TRX_STATE_RX_ON);
+    while(ieee802154_radio_confirm_set_trx_state(dev) == -EAGAIN) {};
```
```suggestion
    while (ieee802154_radio_confirm_set_trx_state(dev) == -EAGAIN) {};
```

jia200x replied:
done!

benpicco reviewed (commented):
looks good, just some style nits

benpicco commented:
Murdock is not happy

jia200x commented:
fixed! I also piggybacked another fix. I was calling one of the `ieee802154_xxx` functions inside an `assert`. This wouldn't work if DEVELHELP is disabled

jia200x commented:
ping! I would like to move on with https://github.com/RIOT-OS/RIOT/pull/16535, but this one is needed :)
